src/compiler/checker.ts

```diff
@@ -33363,10 +33363,7 @@ namespace ts {
 
         function isValidUnusedLocalDeclaration(declaration: Declaration): boolean {
             if (isBindingElement(declaration) && isIdentifierThatStartsWithUnderscore(declaration.name)) {
-                return !!findAncestor(declaration.parent, ancestor =>
-                    isArrayBindingPattern(ancestor) || isVariableDeclaration(ancestor) || isVariableDeclarationList(ancestor) ? false :
-                        isForOfStatement(ancestor) ? true : "quit"
-                );
+                return isArrayBindingPattern(declaration.parent);

```

Comment:
Depending on how you interpret the originating bug, this either needs to allow

1. underscore-prefixes on array binding elements.
-OR-
2. underscore-prefixes everywhere.

If (1), then shouldn't the change be just to move `isArrayBindingPattern` down to the `true` branch?
If (2), then shouldn't the change be to simplify this function considerably, and start it with 
`if (isIdentifierThatStartsWithUnderscore(declaration.name)) return true`?

The current implementation looks like it will break nested array binding elements and for-of destructuring:

```ts
const { x: { _a, _b } } = o
for (const { _a, _b } of os) console.log()
```

And I didn't see any tests for that, so I suspect our test coverage just misses it.

Reply:
@sandersn Thanks. Hm. I thought about the first option., Based on the description of the issue, the main point to allow using `_` in ArrayBinding anywhere, not just in the context of `For / For-Of`. 

> If (1), then shouldn't the change be just to move isArrayBindingPattern down to the true branch?

Not sure that I understood your point :).

> The current implementation looks like it will break nested array binding elements and for-of restructuring:

Do you mean the following use case?
```ts
for (const [[a, _b]] of [[]]) {}
```

> And I didn't see any tests for that, so I suspect our test coverage just misses it.

I think it is not allowed. Do we need to allow it?

Reply:
@sandersn I've changed tests that cover nested array binding elements.

Reply:
My point is that the original code allowed underscore-prefixed binding elements if they were (1) an object element (2) inside a for-of, and nowhere else.

If you just want to extend that to array binding patterns, then underscore-prefixed object binding elements still need to be allowed, at least in for-of, but preferably everywhere. The current change disallows them -- it only returns true for `isArrayBindingPattern(declaration.parent)`.

I think the two object binding examples I gave above will have errors with this change, and that the new change just needs to be `return true` to make those errors go away.

Reply:
@sandersn I suppose we need to allow object element with underscore-prefixed alias 
```ts
const { a /*error*/, _b /*error*/ } = { a: 1, _b: 1 };
const { a: _a /*ok*/, _b: _a /*ok*/ } = { a: 1, _b: 1 };
```
Or 
```ts
const { a /*error*/, _b /*ok*/ } = { a: 1, _b: 1 };
```
?

Reply:
Ah, I didn't consider the `const { _b }  = instance` case as possibly destructuring a 'private' property rather than an intentionally unused one. I think the behaviour you implemented is the right one.